File: 1.5.js
/*
1.5) One Away: There are three types of edits to a given string: insert a character,
 remove a character, or replace a character.
Given two strings, write a function to check if they are one edit away (or zero).

I: Two strings
O: Boolean
C: Time/Space => None.  However, it doesn't hurt to offer the optimal solution.
E:

Same String
Strings of different length
What if a string is length 1? Length 0?
I.e. what if they're both the same length (for 1 or 0)?
Is it one edit per string or one total edit?
*/

var oneAway = (str1, str2) => {
  if (str1 === str2) {
    return true;
  } else if (Math.abs(str1.length - str2.length) > 1) {
    return false;
  } else {
    var hashitout = (str) => str.split('').reduce((a, b) => {
      a[b] = (a[b] || 0) + 1;
      return a;
    }, {});
    var hashOne = hashitout(str1);
    var hashTwo = hashitout(str2);
    var booboo = 0;

    for (var key in hashTwo) {
      if (!(key in hashOne) || (hashOne[key] !== hashTwo[key])) {
        booboo++;
      }

      if (booboo > 1) {
        return false;
      }
    }

    return true;
  }
};

/// WRITE ALL PERMUTATIONS / CONSIDER ALL PERMUTATIONS
// DO NOT JUMP INTO CODING
// PSEUDOCODE
// WRITE DIAGRAMS ETC.
// BE PATIENT
// consider all possibilities/edge cases
// i would try to not erase IOCE, but other stuff should be safe

File: 1.5.go
/*
1.5) One Away: There are three types of edits to a given string: insert a character,
 remove a character, or replace a character.
Given two strings, write a function to check if they are one edit away (or zero).

I: Two strings
O: Boolean
C: Time/Space => None.  However, it doesn't hurt to offer the optimal solution.
E:

Are they the same string?
If so, return true.

Space?
Case sensitivity?
Empty string and a letter?
*/

package main

import (
  "fmt"
  "math"
)

func OneAway (a, b string) bool {
  if a == b {
    return true
    // if they're the same string, it's true!
  } else if math.Abs(float64(len(a) - len(b))) > 1 {
    // if the difference in length is greater than one, removing, adding or
    // replacing one character will not suffice, so it must be false
    return false
  } else {
    runeHashA := map[rune]int{}
    runeHashB := map[rune]int{}
    // to initiate a map, need to either use make OR
    // type inference assignment with curly braces

    for _, runevalue := range a {
      // range on a string offers the index and the RUNE value
	     if runeHashA[runevalue] == 0 {
         runeHashA[runevalue] = 1
       } else {
         runeHashA[runevalue] += 1
       }
    }
    for _, runevalueb := range b {
      // range on a string offers the index and the RUNE value
       if runeHashB[runevalueb] == 0 {
         runeHashB[runevalueb] = 1
       } else {
         runeHashB[runevalueb] += 1
       }
    }

    flawcounter := 0
    for k, _ := range runeHashA {
      if runeHashB[k] == 0 || runeHashA[k] != runeHashB[k] {
        flawcounter++
      }
      if flawcounter > 1 {
        return false
      }
    }
    return true

  }
}
func main() {
  fmt.Println(OneAway("dag", "dog"))
}

// Time: O(n) Space: O(n)
